libvmi/driver/kvm/kvm.c

```diff
@@ -547,40 +165,190 @@ status_t
+        /* From this point, kvm->kvmi_dom won't be touched. */
+        kvm->libkvmi.kvmi_domain_close(kvm->kvmi_dom, true);
+        return false;
+    }
+
+    return true;
+}
+
+static bool
+get_kvmi_registers(
+    kvm_instance_t *kvm,
+    reg_t reg,
+    uint64_t *value)
+{
+    struct kvm_regs regs;
+    struct kvm_sregs sregs;
+    struct {
+        struct kvm_msrs      msrs;
+        struct kvm_msr_entry entries[2];
+    } msrs = {0};
+    unsigned int mode;
+    unsigned short vcpu = 0;
+    int err;
+
+    if (!kvm->kvmi_dom)
+        return false;
+
+    msrs.msrs.nmsrs = sizeof(msrs.entries)/sizeof(msrs.entries[0]);
+    msrs.entries[0].index = msr_index[MSR_EFER];
+    msrs.entries[1].index = msr_index[MSR_STAR];
+
+    err = kvm->libkvmi.kvmi_get_registers(kvm->kvmi_dom, vcpu, &regs, &sregs, &msrs.msrs, &mode);
+
+    if (err != 0)
+        return false;
+
+    /* mode should be 8 if VMI_PM_IA32E == vmi->page_mode */
+
+    switch (reg) {
+        case RAX:
+            *value = regs.rax;
+            break;
+        case RBX:
+            *value = regs.rbx;
+            break;
+        case RCX:
+            *value = regs.rcx;
+            break;
+        case RDX:
+            *value = regs.rdx;
+            break;
+        case RBP:
+            *value = regs.rbp;
+            break;
+        case RSI:
+            *value = regs.rsi;
+            break;
+        case RDI:
+            *value = regs.rdi;
+            break;
+        case RSP:
+            *value = regs.rsp;
+            break;
+        case R8:
+            *value = regs.r8;
+            break;
+        case R9:
+            *value = regs.r9;
+            break;
+        case R10:
+            *value = regs.r10;
+            break;
+        case R11:
+            *value = regs.r11;
+            break;
+        case R12:
+            *value = regs.r12;
+            break;
+        case R13:
+            *value = regs.r13;
+            break;
+        case R14:
+            *value = regs.r14;
+            break;
+        case R15:
+            *value = regs.r15;
+            break;
+        case RIP:
+            *value = regs.rip;
+            break;
+        case RFLAGS:
+            *value = regs.rflags;
+            break;
+        case CR0:
+            *value = sregs.cr0;
+            break;
+        case CR2:
+            *value = sregs.cr2;
+            break;
+        case CR3:
+            *value = sregs.cr3;
+            break;
+        case CR4:
+            *value = sregs.cr4;
+            break;
+        case FS_BASE:
+            *value = sregs.fs.base;
+            break;
+        case GS_BASE:
+            *value = sregs.gs.base;
+            break;
+        case MSR_EFER:
+            *value = msrs.entries[0].data;
+            break;
+        case MSR_STAR:
+            *value = msrs.entries[1].data;
+            break;
+        default:
+            dbprint(VMI_DEBUG_KVM, "--Reading register %"PRIu64" not implemented\n", reg);
```

Comment:
Looks like you have MSR_LSTAR implemented in kvm_get_vcpuregs but not here

Reply:
Good catch.
I also added `MSR_IA32_SYSENTER_CS/ESP/EIP`.